File: calculations/run_all.m
function run_all
%RUN_ALL Reproduce the first numerical study for the manuscript.
% Uses only core MATLAB functions and writes all results under results/.

root = fileparts(fileparts(mfilename('fullpath')));
out = fullfile(root, 'results');
if ~exist(out, 'dir'), mkdir(out); end

mat.name = 'Aluminum (representative isotropic values)';
mat.rho = 2700;            % kg/m^3
mat.cL = 6320;             % m/s
mat.cT = 3130;             % m/s
mat.d = 1.0e-3;            % total plate thickness, m
mat.h = mat.d/2;

opts.K = linspace(0.04, 8.0, 260).';       % dimensionless kh
opts.Wmax = 12.0;                          % omega*h/cT
opts.scanPoints = 7000;
opts.nBranches = 3;

fprintf('Solving Rayleigh-Lamb equations...\n');
[A, Ares] = solve_family('A', mat, opts);
[S, Sres] = solve_family('S', mat, opts);

% Dimensional quantities and derivatives.
A = add_kinematics(A, opts.K, mat);
S = add_kinematics(S, opts.K, mat);

% Low-frequency validation.
E = mat.rho*mat.cT^2*(3*mat.cL^2-4*mat.cT^2)/(mat.cL^2-mat.cT^2);
nu = (mat.cL^2-2*mat.cT^2)/(2*(mat.cL^2-mat.cT^2));
cPlate = sqrt(E/(mat.rho*(1-nu^2)));
Dplate = E*mat.d^3/(12*(1-nu^2));
kSmall = opts.K(1:15)/mat.h;
S0expected = cPlate*kSmall;
A0expected = sqrt(Dplate/(mat.rho*mat.d))*kSmall.^2;
S0err = median(abs(S.omega(1:15,1)-S0expected)./S0expected);
A0err = median(abs(A.omega(1:15,1)-A0expected)./A0expected);

% ZGV candidates are local frequency minima away from endpoints.
zgvA = find_zgv(A, opts.K, mat, 'A');
zgvS = find_zgv(S, opts.K, mat, 'S');

% Causal reduced-order calculation on the fundamental branches.
fprintf('Evaluating causal modal integrals...\n');
transient = causal_study(A, S, opts.K, mat);

% Tables and plots.
write_dispersion_csv(fullfile(out, 'dispersion.csv'), opts.K, A, S, mat);
plot_dispersion(fullfile(out, 'dispersion.png'), A, S, mat);
plot_group_velocity(fullfile(out, 'group_velocity.png'), A, S, mat);
plot_transient(fullfile(out, 'causal_buildup.png'), transient);
write_summary(fullfile(out, 'validation_summary.txt'), mat, opts, Ares, Sres, ...
    S0err, A0err, cPlate, zgvA, zgvS, transient);

save(fullfile(out, 'calculation_data.mat'), 'mat', 'opts', 'A', 'S', ...
    'Ares', 'Sres', 'zgvA', 'zgvS', 'transient');
fprintf('Finished. Results written to %s\n', out);
end

function [branches, maxResidual] = solve_family(family, mat, opts)
K = opts.K;
Wgrid = linspace(1e-4, opts.Wmax, opts.scanPoints);
W = nan(numel(K), opts.nBranches);
residuals = nan(size(W));
for i = 1:numel(K)
    vals = arrayfun(@(w) characteristic(w, K(i), mat.cT/mat.cL, family), Wgrid);
    roots = [];
    for j = 1:numel(Wgrid)-1
        if near_bulk_threshold(Wgrid(j), K(i), mat.cT/mat.cL), continue; end
        if isfinite(vals(j)) && isfinite(vals(j+1)) && vals(j)*vals(j+1) < 0
            try
                r = fzero(@(w) characteristic(w, K(i), mat.cT/mat.cL, family), ...
                    [Wgrid(j), Wgrid(j+1)]);
                if ~near_bulk_threshold(r, K(i), mat.cT/mat.cL) && ...
                        (isempty(roots) || min(abs(roots-r)) > 2e-4)
                    roots(end+1) = r; %#ok<AGROW>
                end
            catch
            end
        end
    end
    roots = sort(roots);
    take = min(opts.nBranches, numel(roots));
    if take > 0
        W(i,1:take) = roots(1:take);
        for b = 1:take
            residuals(i,b) = normalized_residual(roots(b), K(i), mat.cT/mat.cL, family);
        end
    end
end
branches.W = W;
branches.omega = W*mat.cT/mat.h;
maxResidual = max(residuals(:), [], 'omitnan');
end

function tf = near_bulk_threshold(W, K, r)
tf = abs(W-K) < 5e-3 || abs(W-K/r) < 5e-3;
end

function y = characteristic(W, K, r, family)
p = sqrt(complex((r*W)^2-K^2));
q = sqrt(complex(W^2-K^2));
if family == 'S'
    detv = (q^2-K^2)^2*cos(p)*sin(q) + 4*K^2*p*q*sin(p)*cos(q);
else
    detv = 4*K^2*p*q*cos(p)*sin(q) + (q^2-K^2)^2*sin(p)*cos(q);
end
if W < K
    y = imag(detv);
else
    y = real(detv);
end
scale = 1 + abs((q^2-K^2)^2) + abs(4*K^2*p*q);
y = y/scale;
end

function r = normalized_residual(W, K, ratio, family)
r = abs(characteristic(W, K, ratio, family));
end

function B = add_kinematics(B, K, mat)
k = K/mat.h;
B.k = k;
for b = 1:size(B.omega,2)
    good = isfinite(B.omega(:,b));
    B.vg(:,b) = nan(size(k));
    B.curvature(:,b) = nan(size(k));
    if nnz(good) > 4
        B.vg(good,b) = gradient(B.omega(good,b), k(good));
        B.curvature(good,b) = gradient(B.vg(good,b), k(good));
    end
end
end

function z = find_zgv(B, K, mat, family)
z = struct('mode', {}, 'K', {}, 'fd_MHz_mm', {}, 'frequency_MHz', {});
for b = 2:size(B.W,2) % exclude fundamental acoustic branch
    w = B.W(:,b);
    for i = 3:numel(K)-2
        if all(isfinite(w(i-1:i+1))) && w(i) < w(i-1) && w(i) < w(i+1)
            n = numel(z)+1;
            z(n).mode = sprintf('%s%d', family, b-1);
            z(n).K = K(i);
            z(n).fd_MHz_mm = B.omega(i,b)/(2*pi)*mat.d/1000;
            z(n).frequency_MHz = B.omega(i,b)/(2*pi)/1e6;
        end
    end
end
end

function R = causal_study(A, S, K, mat)
% Normalized modal coupling: Gaussian line source and scalar modal response.
targetK = 1.0;
[~, idx] = min(abs(K-targetK));
Omega = A.omega(idx,1);
w0 = 1.0e-3;
x = 20e-3;
gamma = 0.01*Omega;
k = A.k;
mask = isfinite(A.omega(:,1)) & isfinite(S.omega(:,1));
k = k(mask); Kuse = K(mask);
wa = A.omega(mask,1); ws = S.omega(mask,1);
source = exp(-(k*w0).^2/4);
dk = mean(diff(k));
weights = source.*cos(k*x)*dk/pi;

T = 2*pi/Omega;
t = linspace(0, 250*T, 2501).';
[ua, usa, utra] = modal_integral(wa, weights, Omega, gamma, t);
[us, uss, utrs] = modal_integral(ws, weights, Omega, gamma, t);
ratioA = abs(utra)/max(abs(usa), eps);
futureMax = flipud(cummax(flipud(ratioA)));
j = find(futureMax < 0.05, 1, 'first');
if isempty(j), tss = NaN; else, tss = t(j); end

vgTarget = A.vg(idx,1);
tArrival = x/abs(vgTarget);
N95 = ceil(tss/T);
purityA = abs(ua).^2./(abs(ua).^2+abs(us).^2+eps);

R.t = t; R.T = T; R.frequency = Omega/(2*pi); R.Omega = Omega;
R.uA = ua; R.uS = us; R.transientA = utra; R.ssA = usa;
R.ratioA = ratioA; R.purityA = purityA;
R.tss = tss; R.tArrival = tArrival; R.N95 = N95;
R.x = x; R.w0 = w0; R.gamma = gamma;
R.targetK = Kuse(find(mask(idx:end),1,'first')); %#ok<FNDSB>
end

function [u, ussTime, utr] = modal_integral(wn, weights, Omega, gamma, t)
H = weights./(wn.^2-Omega^2-2i*gamma*Omega);
wd = sqrt(max(wn.^2-gamma^2, 0));
sp = -gamma+1i*wd;
sm = -gamma-1i*wd;
Cp = H.*(1i*Omega+sm)./(sp-sm);
Cm = -H-Cp;
ssPhasor = sum(H);
ussTime = ssPhasor*exp(-1i*Omega*t);
utr = zeros(size(t));
for j = 1:numel(wn)
    utr = utr + Cp(j)*exp(sp(j)*t) + Cm(j)*exp(sm(j)*t);
end
u = ussTime+utr;
end

function write_dispersion_csv(path, K, A, S, mat)
fid = fopen(path, 'w');
fprintf(fid, 'K');
for b=1:3, fprintf(fid, ',A%d_fd_MHz_mm,A%d_vg_m_s',b-1,b-1); end
for b=1:3, fprintf(fid, ',S%d_fd_MHz_mm,S%d_vg_m_s',b-1,b-1); end
fprintf(fid, '\n');
for i=1:numel(K)
    fprintf(fid, '%.9g', K(i));
    for b=1:3, fprintf(fid, ',%.9g,%.9g', A.omega(i,b)/(2*pi)*mat.d/1000, A.vg(i,b)); end
    for b=1:3, fprintf(fid, ',%.9g,%.9g', S.omega(i,b)/(2*pi)*mat.d/1000, S.vg(i,b)); end
    fprintf(fid, '\n');
end
fclose(fid);
end

function plot_dispersion(path, A, S, mat)
f = figure('Visible','off','Color','w','Position',[100 100 900 560]); hold on;
colors = lines(3);
for b=1:3
    plot(A.k*mat.d, A.omega(:,b)/(2*pi)*mat.d/1000, '--', 'Color',colors(b,:), 'LineWidth',1.5);
    plot(S.k*mat.d, S.omega(:,b)/(2*pi)*mat.d/1000, '-', 'Color',colors(b,:), 'LineWidth',1.5);
end
xlabel('k d'); ylabel('f d (MHz mm)'); grid on; box on;
legend('A_0','S_0','A_1','S_1','A_2','S_2','Location','northwest');
title('Rayleigh-Lamb dispersion: 1 mm isotropic aluminum plate');
exportgraphics(f,path,'Resolution',180); close(f);
end

function plot_group_velocity(path, A, S, mat)
f = figure('Visible','off','Color','w','Position',[100 100 900 560]); hold on;
plot(A.omega(:,1)/(2*pi)*mat.d/1000,A.vg(:,1),'-','LineWidth',1.6);
plot(S.omega(:,1)/(2*pi)*mat.d/1000,S.vg(:,1),'-','LineWidth',1.6);
yline(0,'k:'); xlabel('f d (MHz mm)'); ylabel('group velocity (m/s)');
legend('A_0','S_0','Location','best'); grid on; box on;
title('Fundamental-mode group velocity'); exportgraphics(f,path,'Resolution',180); close(f);
end

function plot_transient(path, R)
f = figure('Visible','off','Color','w','Position',[100 100 900 700]);
tcycles=R.t/R.T;
subplot(2,1,1); plot(tcycles,abs(R.uA)/max(abs(R.ssA)),'LineWidth',1.3); hold on;
yline(1,'k:'); xlabel('time (drive cycles)'); ylabel('|u_{A0}|/|u_{ss}|'); grid on;
title('Normalized causal buildup (scalar modal coupling)');
subplot(2,1,2); plot(tcycles,R.purityA,'LineWidth',1.3);
xlabel('time (drive cycles)'); ylabel('A_0 modal purity'); ylim([0 1]); grid on;
exportgraphics(f,path,'Resolution',180); close(f);
end

function write_summary(path, mat, opts, Ares, Sres, S0err, A0err, cPlate, zA, zS, R)
fid=fopen(path,'w');
fprintf(fid,'CALCULATION AND VALIDATION SUMMARY\n');
fprintf(fid,'Material: %s\n',mat.name);
fprintf(fid,'rho=%.1f kg/m^3, cL=%.1f m/s, cT=%.1f m/s, d=%.6g m\n',mat.rho,mat.cL,mat.cT,mat.d);
fprintf(fid,'Grid: %d K values, %d frequency scan points, Wmax=%.2f\n\n',numel(opts.K),opts.scanPoints,opts.Wmax);
fprintf(fid,'Maximum normalized characteristic residual: A %.3e, S %.3e\n',Ares,Sres);
fprintf(fid,'Low-frequency median relative error: A0 thin-plate %.3f%%, S0 plate-speed %.3f%%\n',100*A0err,100*S0err);
fprintf(fid,'Expected low-frequency S0 plate speed: %.3f m/s\n\n',cPlate);
fprintf(fid,'Detected higher-mode local minima (grid estimates):\n');
allz=[zA,zS];
if isempty(allz), fprintf(fid,'  none on sampled branches/range\n'); end
for i=1:numel(allz)
    fprintf(fid,'  %s: K=%.4f, f*d=%.5f MHz mm, f=%.5f MHz\n',allz(i).mode,allz(i).K,allz(i).fd_MHz_mm,allz(i).frequency_MHz);
end
fprintf(fid,'\nNormalized causal study (not an absolute thermoelastic amplitude):\n');
fprintf(fid,'  drive f=%.6f MHz, x=%.3f mm, Gaussian width=%.3f mm, gamma/Omega=%.4f\n',R.frequency/1e6,R.x*1e3,R.w0*1e3,R.gamma/R.Omega);
fprintf(fid,'  group-delay estimate x/|vg|=%.6g us\n',R.tArrival*1e6);
fprintf(fid,'  5%% envelope settling time=%.6g us (%.1f cycles)\n',R.tss*1e6,R.tss/R.T);
fprintf(fid,'  N95 proxy (cycles to remain within 5%%)=%.0f\n',R.N95);
fprintf(fid,'\nValidation scope:\n');
fprintf(fid,'  Dispersion results solve the full isotropic Rayleigh-Lamb equations.\n');
fprintf(fid,'  The causal calculation uses unit scalar modal overlap and constant damping; it validates the computational framework, not an experimental absolute amplitude.\n');
fclose(fid);
end

File: calculations/run_all.py
"""Reproduce the first Lamb-wave dispersion and causal-buildup study."""
from __future__ import annotations

import csv
import os
from dataclasses import dataclass
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
os.environ.setdefault("MPLCONFIGDIR", str(ROOT / ".matplotlib"))

import matplotlib.pyplot as plt
import numpy as np
from scipy.optimize import brentq


OUT = ROOT / "results"
OUT.mkdir(exist_ok=True)


@dataclass(frozen=True)
class Material:
    rho: float = 2700.0
    cL: float = 6320.0
    cT: float = 3130.0
    d: float = 1.0e-3

    @property
    def h(self) -> float:
        return self.d / 2


MAT = Material()


def characteristic(W: float | np.ndarray, K: float, family: str) -> np.ndarray:
    """Scaled, real Rayleigh--Lamb characteristic function."""
    W = np.asarray(W)
    ratio = MAT.cT / MAT.cL
    p = np.sqrt((ratio * W) ** 2 - K**2 + 0j)
    q = np.sqrt(W**2 - K**2 + 0j)
    if family == "S":
        det = (q**2 - K**2) ** 2 * np.cos(p) * np.sin(q)
        det += 4 * K**2 * p * q * np.sin(p) * np.cos(q)
    else:
        det = 4 * K**2 * p * q * np.cos(p) * np.sin(q)
        det += (q**2 - K**2) ** 2 * np.sin(p) * np.cos(q)
    scale = 1 + np.abs((q**2 - K**2) ** 2) + np.abs(4 * K**2 * p * q)
    return np.where(W < K, det.imag, det.real) / scale


def near_threshold(W: float, K: float) -> bool:
    r = MAT.cT / MAT.cL
    # The determinant form contains removable p=0 and q=0 factors. A tolerance
    # wider than either scan spacing prevents those bulk-wave thresholds from
    # being misidentified as Lamb roots under grid refinement.
    return abs(W - K) < 5e-3 or abs(W - K / r) < 5e-3


def solve_family(family: str, K: np.ndarray, wmax=12.0, nscan=7000, nbranches=3):
    grid = np.linspace(1e-4, wmax, nscan)
    result = np.full((K.size, nbranches), np.nan)
    residual = np.full_like(result, np.nan)
    for i, kval in enumerate(K):
        values = characteristic(grid, kval, family)
        crossings = np.flatnonzero(values[:-1] * values[1:] < 0)
        roots: list[float] = []
        for j in crossings:
            if near_threshold(grid[j], kval):
                continue
            try:
                root = brentq(lambda w: float(characteristic(w, kval, family)),
                              grid[j], grid[j + 1], xtol=1e-11)
            except ValueError:
                continue
            if near_threshold(root, kval) or any(abs(root - x) < 2e-4 for x in roots):
                continue
            roots.append(root)
        roots.sort()
        for branch, root in enumerate(roots[:nbranches]):
            result[i, branch] = root
            residual[i, branch] = abs(characteristic(root, kval, family))
    return result, residual


def kinematics(W: np.ndarray, K: np.ndarray):
    omega = W * MAT.cT / MAT.h
    k = K / MAT.h
    vg = np.full_like(omega, np.nan)
    curvature = np.full_like(omega, np.nan)
    for b in range(W.shape[1]):
        good = np.isfinite(omega[:, b])
        vg[good, b] = np.gradient(omega[good, b], k[good])
        curvature[good, b] = np.gradient(vg[good, b], k[good])
    return omega, k, vg, curvature


def local_minima(omega: np.ndarray, K: np.ndarray, family: str, max_k=1.5):
    found = []
    for b in range(1, omega.shape[1]):
        y = omega[:, b]
        for i in range(2, K.size - 2):
            if K[i] <= max_k and np.all(np.isfinite(y[i - 1:i + 2])) and y[i] < y[i - 1] and y[i] < y[i + 1]:
                # Quadratic refinement using the three neighboring grid points.
                coeff = np.polyfit(K[i - 1:i + 2], y[i - 1:i + 2], 2)
                kr = -coeff[1] / (2 * coeff[0])
                wr = np.polyval(coeff, kr)
                found.append((f"{family}{b}", kr, wr / (2*np.pi)*MAT.d/1000, wr/(2*np.pi)/1e6))
    return found


def modal_integral(wn, weights, Omega, gamma, t):
    H = weights / (wn**2 - Omega**2 - 2j*gamma*Omega)
    # Complex square root handles both underdamped and overdamped spectral
    # components without collapsing the two characteristic roots.
    wd = np.sqrt(wn**2 - gamma**2 + 0j)
    sp, sm = -gamma + 1j*wd, -gamma - 1j*wd
    Cp = H * (1j*Omega + sm) / (sp - sm)
    Cm = -H - Cp
    ss = H.sum() * np.exp(-1j*Omega*t)
    transient = (np.exp(np.outer(t, sp)) @ Cp + np.exp(np.outer(t, sm)) @ Cm)
    return ss + transient, ss, transient


def break_plot_jumps(y: np.ndarray, threshold=0.35) -> np.ndarray:
    """Insert NaNs around root-index discontinuities for honest plotting."""
    clean = y.copy()
    jumps = np.flatnonzero(np.abs(np.diff(clean)) > threshold)
    clean[jumps] = np.nan
    clean[jumps + 1] = np.nan
    return clean


def causal_study(K, k, Aomega, Somega, Avg, x=20e-3, width=1e-3,
                 damping_ratio=0.01, ntime=2501):
    target = np.nanargmin(abs(K - 1.0))
    Omega = Aomega[target, 0]
    gamma = damping_ratio * Omega
    # Isolate the driven modal band. Very-low-frequency off-resonant components
    # represent a quasistatic turn-on field and should not be included in the
    # settling metric of the frequency-filtered Lamb-wave signal.
    good_a = np.isfinite(Aomega[:, 0]) & (Aomega[:, 0] > 0.5*Omega) & (Aomega[:, 0] < 1.5*Omega)
    good_s = np.isfinite(Somega[:, 0]) & (Somega[:, 0] > 0.5*Omega) & (Somega[:, 0] < 1.5*Omega)
    ka, ks = k[good_a], k[good_s]
    weights_a = np.exp(-(ka*width)**2/4) * np.cos(ka*x) * np.mean(np.diff(k))/np.pi
    weights_s = np.exp(-(ks*width)**2/4) * np.cos(ks*x) * np.mean(np.diff(k))/np.pi
    period = 2*np.pi/Omega
    t = np.linspace(0, 250*period, ntime)
    ua, ssa, tra = modal_integral(Aomega[good_a, 0], weights_a, Omega, gamma, t)
    us, _, _ = modal_integral(Somega[good_s, 0], weights_s, Omega, gamma, t)
    ratio = abs(tra) / max(abs(ssa[0]), np.finfo(float).eps)
    future_max = np.maximum.accumulate(ratio[::-1])[::-1]
    settled = np.flatnonzero(future_max < 0.05)
    tss = t[settled[0]] if settled.size else np.nan
    purity = abs(ua)**2/(abs(ua)**2 + abs(us)**2 + np.finfo(float).eps)
    return dict(t=t, period=period, frequency=Omega/(2*np.pi), Omega=Omega,
                ua=ua, us=us, ssa=ssa, tra=tra, ratio=ratio, purity=purity,
                tss=tss, arrival=x/abs(Avg[target, 0]), N95=np.ceil(tss/period),
                x=x, width=width, gamma=gamma)


def main():
    K = np.linspace(0.04, 8.0, 260)
    print("Solving antisymmetric branches...")
    AW, Ares = solve_family("A", K)
    print("Solving symmetric branches...")
    SW, Sres = solve_family("S", K)
    Aomega, k, Avg, Acurv = kinematics(AW, K)
    Somega, _, Svg, Scurv = kinematics(SW, K)

    # Thin-plate validation over the smallest sampled K values.
    ratio = MAT.cT / MAT.cL
    E = MAT.rho*MAT.cT**2*(3*MAT.cL**2 - 4*MAT.cT**2)/(MAT.cL**2 - MAT.cT**2)
    nu = (MAT.cL**2 - 2*MAT.cT**2)/(2*(MAT.cL**2 - MAT.cT**2))
    cplate = np.sqrt(E/(MAT.rho*(1-nu**2)))
    rigidity = E*MAT.d**3/(12*(1-nu**2))
    Klow = np.linspace(0.002, 0.08, 20)
    AWlow, _ = solve_family("A", Klow, wmax=0.5, nscan=4000, nbranches=1)
    SWlow, _ = solve_family("S", Klow, wmax=0.5, nscan=4000, nbranches=1)
    Aolow, klow, _, _ = kinematics(AWlow, Klow)
    Solow, _, _, _ = kinematics(SWlow, Klow)
    s_expected = cplate*klow
    a_expected = np.sqrt(rigidity/(MAT.rho*MAT.d))*klow**2
    s_err = np.nanmedian(abs(Solow[:, 0]-s_expected)/s_expected)
    a_err = np.nanmedian(abs(Aolow[:, 0]-a_expected)/a_expected)
    zgvs = local_minima(Aomega, K, "A") + local_minima(Somega, K, "S")
    transient = causal_study(K, k, Aomega, Somega, Avg)

    # Convergence checks: repeat the characteristic scan at half resolution,
    # and repeat the transient calculation with half the time step.
    AWcoarse, _ = solve_family("A", K, nscan=3500)
    SWcoarse, _ = solve_family("S", K, nscan=3500)
    dispersion_convergence = np.nanmax(np.abs(np.r_[AWcoarse-AW, SWcoarse-SW]))
    transient_fine = causal_study(K, k, Aomega, Somega, Avg, ntime=5001)
    time_convergence_cycles = abs(transient_fine["tss"]-transient["tss"])/transient["period"]

    # Verify the analytic oscillator constants enforce zero initial conditions.
    test_wn, test_Omega, test_gamma = 1.2, 1.0, 0.01
    H = 1/(test_wn**2-test_Omega**2-2j*test_gamma*test_Omega)
    wd = np.sqrt(test_wn**2-test_gamma**2+0j)
    sp, sm = -test_gamma+1j*wd, -test_gamma-1j*wd
    Cp = H*(1j*test_Omega+sm)/(sp-sm); Cm = -H-Cp
    initial_displacement_error = abs(H+Cp+Cm)
    initial_velocity_error = abs(-1j*test_Omega*H+sp*Cp+sm*Cm)

    # Parameter map for the design observables requested in the manuscript.
    sweep = []
    for distance in (10e-3, 20e-3, 40e-3):
        for width in (0.5e-3, 1e-3, 2e-3):
            for damping in (0.005, 0.01, 0.02):
                q = causal_study(K, k, Aomega, Somega, Avg, x=distance,
                                 width=width, damping_ratio=damping, ntime=2501)
                sweep.append((distance*1e3, width*1e3, damping,
                              q["arrival"]*1e6, q["tss"]*1e6, q["N95"],
                              np.mean(q["purity"][-100:])))

    # Machine-readable output.
    headers = ["K"]
    columns = [K]
    for family, omega, vg in (("A", Aomega, Avg), ("S", Somega, Svg)):
        for b in range(3):
            headers += [f"{family}{b}_fd_MHz_mm", f"{family}{b}_vg_m_s"]
            columns += [omega[:, b]/(2*np.pi)*MAT.d/1000, vg[:, b]]
    with (OUT/"dispersion.csv").open("w", newline="") as f:
        writer = csv.writer(f); writer.writerow(headers); writer.writerows(zip(*columns))
    with (OUT/"parameter_sweep.csv").open("w", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(["distance_mm", "gaussian_width_mm", "gamma_over_omega",
                         "group_delay_us", "settling_time_us", "N95_cycles",
                         "late_time_A0_purity"])
        writer.writerows(sweep)
    np.savez_compressed(OUT/"calculation_data.npz", K=K, k=k, AW=AW, SW=SW,
                        Aomega=Aomega, Somega=Somega, Avg=Avg, Svg=Svg,
                        Acurvature=Acurv, Scurvature=Scurv, **transient)

    # Figures.
    colors = plt.cm.tab10.colors
    fig, ax = plt.subplots(figsize=(8.2, 5.2))
    for b in range(3):
        ay = break_plot_jumps(Aomega[:, b]/(2*np.pi)*MAT.d/1000)
        sy = break_plot_jumps(Somega[:, b]/(2*np.pi)*MAT.d/1000)
        ax.plot(k*MAT.d, ay, "--", color=colors[b], label=f"A{b}")
        ax.plot(k*MAT.d, sy, "-", color=colors[b], label=f"S{b}")
    ax.set(xlabel=r"$kd$", ylabel=r"$fd$ (MHz mm)", title="Rayleigh–Lamb dispersion: 1 mm aluminum plate")
    ax.grid(True, alpha=.3); ax.legend(ncol=2); fig.tight_layout(); fig.savefig(OUT/"dispersion.png", dpi=180); plt.close(fig)

    fig, ax = plt.subplots(figsize=(8.2, 5.2))
    ax.plot(Aomega[:, 0]/(2*np.pi)*MAT.d/1000, Avg[:, 0], label="A0")
    ax.plot(Somega[:, 0]/(2*np.pi)*MAT.d/1000, Svg[:, 0], label="S0")
    ax.axhline(0, color="k", lw=.8, ls=":"); ax.grid(True, alpha=.3); ax.legend()
    ax.set(xlabel=r"$fd$ (MHz mm)", ylabel="group velocity (m/s)", title="Fundamental-mode group velocity")
    fig.tight_layout(); fig.savefig(OUT/"group_velocity.png", dpi=180); plt.close(fig)

    cycles = transient["t"]/transient["period"]
    fig, axs = plt.subplots(2, 1, figsize=(8.2, 6.5), sharex=True)
    axs[0].plot(cycles, abs(transient["ua"])/max(abs(transient["ssa"])))
    axs[0].axhline(1, color="k", lw=.8, ls=":"); axs[0].set_ylabel(r"$|u_{A0}|/|u_{ss}|$")
    axs[1].plot(cycles, transient["purity"]); axs[1].set(xlabel="time (drive cycles)", ylabel="A0 modal purity", ylim=(0,1))
    for ax in axs: ax.grid(True, alpha=.3)
    fig.suptitle("Normalized causal buildup (scalar modal coupling)"); fig.tight_layout(); fig.savefig(OUT/"causal_buildup.png", dpi=180); plt.close(fig)

    with (OUT/"validation_summary.txt").open("w", encoding="utf-8") as f:
        f.write("CALCULATION AND VALIDATION SUMMARY\n")
        f.write(f"rho={MAT.rho:g} kg/m^3, cL={MAT.cL:g} m/s, cT={MAT.cT:g} m/s, d={MAT.d:g} m\n")
        f.write(f"Grid: {K.size} K values; 7000 frequency samples; Wmax=12\n\n")
        f.write(f"Maximum normalized characteristic residual: A={np.nanmax(Ares):.3e}, S={np.nanmax(Sres):.3e}\n")
        f.write(f"Low-frequency median relative error: A0 thin-plate={100*a_err:.3f}%, S0 plate-speed={100*s_err:.3f}%\n")
        f.write(f"Expected low-frequency S0 plate speed={cplate:.3f} m/s; cT/cL={ratio:.6f}\n\n")
        f.write("Convergence and analytic checks:\n")
        f.write(f"  max |W(3500 scan)-W(7000 scan)|={dispersion_convergence:.3e}\n")
        f.write(f"  settling-time change on halving dt={time_convergence_cycles:.3f} drive cycles\n")
        f.write(f"  oscillator initial displacement residual={initial_displacement_error:.3e}\n")
        f.write(f"  oscillator initial velocity residual={initial_velocity_error:.3e}\n\n")
        f.write("Detected higher-mode local minima (quadratically refined):\n")
        for mode, kval, fd, freq in zgvs:
            f.write(f"  {mode}: K={kval:.5f}, fd={fd:.6f} MHz mm, f={freq:.6f} MHz\n")
        f.write("\nNormalized causal study (not an absolute thermoelastic amplitude):\n")
        f.write(f"  drive={transient['frequency']/1e6:.6f} MHz; x={transient['x']*1e3:.3f} mm; width={transient['width']*1e3:.3f} mm; gamma/Omega=0.01\n")
        f.write(f"  group delay x/|vg|={transient['arrival']*1e6:.6f} us\n")
        f.write(f"  5% settling time={transient['tss']*1e6:.6f} us ({transient['tss']/transient['period']:.1f} cycles)\n")
        f.write(f"  N95 proxy={transient['N95']:.0f} cycles\n\n")
        finite_n = np.array([row[5] for row in sweep], dtype=float)
        f.write(f"Parameter sweep (27 cases): finite N95 range={np.nanmin(finite_n):.0f}--{np.nanmax(finite_n):.0f} cycles; see parameter_sweep.csv\n\n")
        f.write("Dispersion uses the full isotropic Rayleigh–Lamb equations. The causal study uses unit scalar modal overlap and constant damping; it validates the framework, not absolute experimental amplitude.\n")
    print((OUT/"validation_summary.txt").read_text())


if __name__ == "__main__":
    main()
